collator/src/lib.rs

```diff
@@ -296,15 +331,50 @@ where
 				None => 0,
 			};
 
+			let horizontal_messages = sp_io::storage::get(well_known_keys::HRMP_OUTBOUND_MESSAGES);
+			let horizontal_messages = match horizontal_messages
+				.map(|v| Vec::<OutboundHrmpMessage>::decode(&mut &v[..]))
+			{
+				Some(Ok(horizontal_messages)) => horizontal_messages,
+				Some(Err(e)) => {
+					error!(
+						target: "cumulus-collator",
+						"Failed to decode the horizontal messages: {:?}",
+						e
+					);
+					return None
+				}
+				None => Vec::new(),
+			};
+
+			let hrmp_watermark = sp_io::storage::get(well_known_keys::HRMP_WATERMARK);
+			let hrmp_watermark = match hrmp_watermark.map(|v| PBlockNumber::decode(&mut &v[..])) {
+				Some(Ok(hrmp_watermark)) => hrmp_watermark,
+				Some(Err(e)) => {
+					error!(
+						target: "cumulus-collator",
+						"Failed to decode the HRMP watermark: {:?}",
+						e
+					);
+					return None
+				}
+				None => {
+					// If the runtime didn't set `HRMP_WATERMARK`, then it means no messages were
```

Comment:
Should we, as a honest collator do the check here as well?

Reply:
to see if the runtime tries to trick us? Interesting. I haven't thought about that. 

I would say no. The runtime is always more privileged than the node. We build cunning mechanisms to remove the possibility of nodes to bamboozle the runtime. However, I don't see how we could do the other way around: the runtime always can go rogue and sabotage the parachain (or essentially itself), be it subtly (by, say, not reading a message that it must've read) or blatantly (say, by just panicking unconditionally). And there is nothing that the node can do to prevent that.

So I think we do not need an additional check here - it would be superfluous

Reply:
Yeah good point